File: selenium-tests/generate_test_cases.js
const ExcelJS = require('exceljs');

async function generateTestCases() {
    const workbook = new ExcelJS.Workbook();
    workbook.creator = 'Selenium Automated Generator';
    workbook.created = new Date();

    const sheet = workbook.addWorksheet('Web Test Cases');
    
    // Define columns
    sheet.columns = [
        { header: 'Test Case ID', key: 'id', width: 15 },
        { header: 'Module', key: 'module', width: 20 },
        { header: 'Test Scenario', key: 'scenario', width: 40 },
        { header: 'Test Steps', key: 'steps', width: 50 },
        { header: 'Expected Result', key: 'expected', width: 40 },
        { header: 'Priority', key: 'priority', width: 15 },
        { header: 'Status', key: 'status', width: 15 }
    ];

    // Style the header row
    sheet.getRow(1).font = { bold: true };

    const testCases = [];
    let tcId = 1;

    const modules = {
        "Authentication": ["Login valid", "Login invalid", "Empty credentials", "SQL injection in login", "Session timeout", "Logout"],
        "Responsive Design": ["Desktop view rendering", "Mobile view rendering", "Tablet view rendering", "Navbar hamburger toggle", "Image scaling"],
        "Dashboard": ["Data fetch on load", "Mocked API fallback", "Chart rendering", "Widget refresh", "Profile dropdown"],
        "Fitness & Diet": ["Form submission (new meal)", "Input validation (calories)", "List dynamic update", "Delete item", "Filter items"],
        "AI Chat": ["Send valid prompt", "Empty prompt", "WebSockets connection", "Auto-scroll on new message", "Typing indicator"],
        "Cross-Browser": ["Chrome compatibility", "Firefox compatibility", "Edge compatibility", "Safari compatibility"],
        "Security": ["XSS payload in inputs", "CORS policy check", "Content Security Policy (CSP)", "Local storage encryption"],
        "Profile & Settings": ["Update email", "Change password", "Upload avatar", "Delete account", "Toggle theme", "Notification preferences"]
    };

    const edgeCases = [
        "with normal usage",
        "with slow 3G network simulation",
        "with offline mode simulation",
        "with large data payloads",
        "with special characters input",
        "with concurrent requests",
        "with aggressive double-clicking",
        "with expired JWT token"
    ];

    const priorities = ["High", "Medium", "Low", "Critical"];

    // Generate base cases + edge cases combination
    for (const [moduleName, features] of Object.entries(modules)) {
        for (const feature of features) {
            for (const edge of edgeCases) {
                // Determine priority
                let priority = priorities[Math.floor(Math.random() * priorities.length)];
                if (feature.includes("injection") || feature.includes("Security") || edge.includes("token")) {
                    priority = "Critical";
                }

                testCases.push({
                    id: `WEB_TC_${String(tcId).padStart(4, '0')}`,
                    module: moduleName,
                    scenario: `Verify ${feature} ${edge}`,
                    steps: `1. Open web application. 2. Navigate to ${moduleName}. 3. Execute ${feature} ${edge}.`,
                    expected: `Application should handle the scenario appropriately without crashing or exposing sensitive data.`,
                    priority: priority,
                    status: 'Not Executed'
                });
                tcId++;
            }
        }
    }

    // Add E2E flow cases
    const e2eFlows = [
        "Login -> Navigate Dashboard -> Add Meal -> Verify Diet List -> Logout",
        "Signup -> Mobile View -> Open AI Chat -> Prompt -> Receive Response -> Auto-scroll check",
        "Login -> Offline mode -> Try fetching Dashboard -> Verify fallback UI"
    ];

    for (const flow of e2eFlows) {
        testCases.push({
            id: `WEB_TC_${String(tcId).padStart(4, '0')}`,
            module: "E2E Workflows",
            scenario: `Verify complete flow: ${flow}`,
            steps: `Execute steps in sequence: ${flow}`,
            expected: "Flow completes successfully.",
            priority: "High",
            status: "Not Executed"
        });
        tcId++;
    }

    // Add to sheet
    sheet.addRows(testCases);

    const filename = 'Web_Test_Cases.xlsx';
    await workbook.xlsx.writeFile(filename);
    console.log(`Successfully generated ${testCases.length} web test cases in ${filename}`);
}

generateTestCases().catch(err => {
    console.error("Error generating test cases:", err);
});

File: appium-tests/generate_test_cases.py
import openpyxl
import random

def generate_test_cases():
    test_cases = []
    tc_id = 1
    
    modules = {
        "Authentication": ["Login", "Sign Up", "Forgot Password", "Session Timeout", "OAuth Login"],
        "Onboarding": ["Splash Screen", "Permissions Request", "Tutorial Intro"],
        "Dashboard": ["Widget Load", "Navigation Menu", "Pull to Refresh", "Offline mode notification", "User Avatar"],
        "Fitness": ["Add Workout", "Edit Workout", "Delete Workout", "View History", "Sync with smartwatch"],
        "Diet": ["Add Meal", "Search Recipe", "Filter by Cuisine", "View Ingredients", "Calorie calculation"],
        "Budget": ["Set Budget limit", "View Expenses", "Add Expense", "Category breakdown chart"],
        "AI Chat": ["Send prompt", "Receive response", "Network error handling", "Clear chat history", "Voice input"],
        "Profile": ["Update Email", "Change Password", "Upload Profile Picture", "Toggle Dark Mode", "Delete Account"]
    }
    
    edge_cases = [
        "with empty fields",
        "with extremely long string inputs",
        "with special characters",
        "with SQL injection payloads",
        "with XSS payloads",
        "while device is offline",
        "with poor network connection",
        "while switching apps (background to foreground)",
        "with screen rotation (portrait to landscape)",
        "with low battery mode enabled",
        "with rapid multiple clicks",
        "with invalid data formats"
    ]
    
    priorities = ["High", "Medium", "Low"]
    
    # Base cases
    for module, features in modules.items():
        for feature in features:
            test_cases.append({
                "Test Case ID": f"TC_{tc_id:04d}",
                "Module": module,
                "Test Scenario": f"Verify successful {feature}",
                "Test Steps": f"1. Navigate to {module}. 2. Initiate {feature}. 3. Provide valid inputs. 4. Submit.",
                "Expected Result": f"{feature} should execute successfully.",
                "Priority": "High",
                "Status": "Not Executed"
            })
            tc_id += 1
            
            for edge in edge_cases:
                priority = random.choice(priorities)
                if "SQL" in edge or "XSS" in edge:
                    priority = "High"
                
                test_cases.append({
                    "Test Case ID": f"TC_{tc_id:04d}",
                    "Module": module,
                    "Test Scenario": f"Verify {feature} {edge}",
                    "Test Steps": f"1. Navigate to {module}. 2. Initiate {feature} {edge}. 3. Submit.",
                    "Expected Result": f"System should handle the scenario gracefully without crashing. Appropriate error messages if applicable.",
                    "Priority": priority,
                    "Status": "Not Executed"
                })
                tc_id += 1
    
    # E2E Workflows
    e2e_flows = [
        ("Login -> Navigate to Dashboard -> Add Meal -> Verify Calories -> Logout", "High"),
        ("Signup -> Complete Onboarding -> Set Budget -> Add Expense -> Verify Chart", "High"),
        ("Login -> Open AI Chat -> Ask for Recipe -> Add to Diet -> Verify Dashboard", "High")
    ]
    
    for flow, prio in e2e_flows:
        test_cases.append({
            "Test Case ID": f"TC_{tc_id:04d}",
            "Module": "E2E Workflows",
            "Test Scenario": f"Verify E2E flow: {flow}",
            "Test Steps": f"Execute the following flow sequentially: {flow}",
            "Expected Result": "Flow should complete without any blocking issues.",
            "Priority": prio,
            "Status": "Not Executed"
        })
        tc_id += 1
        
    # Create Excel file using openpyxl
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Test Cases"
    
    # Write Headers
    headers = ["Test Case ID", "Module", "Test Scenario", "Test Steps", "Expected Result", "Priority", "Status"]
    ws.append(headers)
    
    # Write Data
    for tc in test_cases:
        row = [tc[h] for h in headers]
        ws.append(row)
        
    filename = "Test_Cases_Summary_and_Details.xlsx"
    wb.save(filename)
    print(f"Successfully generated {len(test_cases)} test cases in {filename}")

if __name__ == "__main__":
    generate_test_cases()
